### PR TITLE
Switch to clap derive API in safekeeper.

### PR DESCRIPTION
**Pull request by arssher**
Less lines and easier to read/modify.

Practically no functional changes.

### CODE REVIEW

**Comment by arssher** [comment]
Based on [#277](https://github.com/neondatabase/neon/pull/2733) PR to less rebase it.

File: safekeeper/src/bin/safekeeper.rs [hunk 1]
```diff
@@ -298,110 +307,21 @@ fn set_id(conf: &mut SafeKeeperConf, given_id: Option<NodeId>) -> Result<()> {
...
-            Arg::new("log-format")
-                .long("log-format")
-                .help("Format for logging, either 'plain' or 'json'")
-        )
+    Ok(my_id)
 }
 
-#[test]
-fn verify_cli() {
-    cli().debug_assert();
```
Comment (SomeoneToIgnore) [inline safekeeper/src/bin/safekeeper.rs:406]: There's a way to debug assert derive declarations too.
See `Migrating` section in https://github.com/clap-rs/clap/releases/tag/v4.0.0-rc.1

```rust 
#[derive(clap::Parser)]
struct Cli {
    ...
}

#[test]
fn verify_cli() {
    use clap::CommandFactory;
    Cli::command().debug_assert()
}
```

It once detected an error where `-h` was specified as a short version of the flag and the help generation of clap shadowed that.
So, could be helpful and seems very easy to create.

**Review by SomeoneToIgnore — COMMENTED** [review]
Looks more readable.
Have you tried running `build --timings` to see how procmacros influence that?

File: safekeeper/src/bin/safekeeper.rs (hunk 1 above)
Comment (arssher) [reply safekeeper/src/bin/safekeeper.rs:406]: Cool, didn't know that. Added.

**Comment by arssher** [comment]
> Have you tried running build --timings to see how procmacros influence that?

I don't think it would be noticeable with other heavy deps, but let's see.

with patch
```

**Comment by arssher** [comment]
> Have you tried running build --timings to see how procmacros influence that?

I don't think it would be noticeable with other heavy deps, but let's see.

with patch
```
cargo build -p safekeeper --timings
Total time: | 466.4s (7m 46.4s)
  | safekeeper v0.1.0 bin "safekeeper" | 14.1s
```

without
```
total 483.2s (8m 3.2s)
safekeeper v0.1.0 bin "safekeeper" | 14.7s
```

and safekeeper v0.1.0 bin "safekeeper" doesn't have code generation part at all.

**Review by SomeoneToIgnore — APPROVED** [review]

**Comment by petuhovskiy** [comment]
GitHub shows a large diff with `+1,225 −2,451`, but it seems that the patch is only a single commit. Not sure if this PR was needed at all, we could review it in the same neon-broker PR.

**Comment by arssher** [comment]
Rebased on main.

File: safekeeper/src/bin/safekeeper.rs [hunk 2]
```diff
@@ -44,136 +47,143 @@ const ID_FILE_NAME: &str = "safekeeper.id";
...
+A fleet of safekeepers is responsible for reliably storing WAL received from
+compute, passing it through consensus (mitigating potential computes brain
+split), and serving the hardened part further downstream to pageserver(s).
+"#;
+
+#[derive(Parser)]
+#[command(name = "Neon safekeeper", version = GIT_VERSION, about = ABOUT, long_about = None)]
+struct Args {
+    /// Path to the safekeeper data directory.
+    #[arg(short = 'D', long, default_value = "./")]
```
Comment (yeputons) [inline safekeeper/src/bin/safekeeper.rs:60]: It was `long = 'dir'` before. This option was never actually parsed by safekeepers. See #3088

**Pull request merged**